doc/specs/stdlib_selection.md

```diff
@@ -0,0 +1,340 @@
+`1:size(array)`, which is modified instead. On return the index array is modified
+such that `all(array(index(1:k)) <= array(index(k)))` and `all(array(k) <= array(k+1:size(array)))`.
+The user can optionally specify `left` and `right` indices to constrain the search
+for the k-th smallest value. This can be useful if you have previously called `arg_select`
+to find a smaller or larger rank (that will have led to partial sorting of
+`index`, thus implying some constraints on the location).
+
+### Licensing
+
+The Fortran Standard Library is distributed under the MIT
+License. It is worth noting that  `select` and `arg_select`
+were derived using some code from quickSelect in the Coretran library, by Leon Foks,
+https://github.com/leonfoks/coretran. Leon Foks has given permission for the code here
+to be released under stdlib's MIT license.
+
+## `select` - find the k-th smallest value in an input array
+
+### Status
+
+Experimental
+
+### Description
+
+Returns the k-th smallest value of `array(:)`, and also partially sorts `array(:)`
+such that `all(array(1:k) <= array(k))`  and `all(array(k) <= array((k+1):size(array)))`
+
+### Syntax
+
+`call [[stdlib_selection(module):select(interface)]]( array, k, kth_smallest [, left, right ] )`
+
+### Class
+
+Generic subroutine.
+
+### Arguments
+
+`array` : shall be a rank one array of any of the types:
+`integer(int8)`, `integer(int16)`, `integer(int32)`, `integer(int64)`,
+`real(sp)`, `real(dp)`, `real(xdp)`, `real(qp)`. It is an `intent(inout)` argument. 
+
+`k`: shall be a scalar with any of the types:
+`integer(int8)`, `integer(int16)`, `integer(int32)`, `integer(int64)`. It
+is an `intent(in)` argument. We search for the `k`-th smallest entry of `array(:)`.
+
+`kth_smallest`: shall be a scalar with the same type as `array`. It is an
+`intent(out)` argument. On return it contains the k-th smallest entry of
+`array(:)`.
+
+`left` (optional): shall be a scalar with the same type as `k`. It is an
+`intent(in)` argument. If specified then we assume the k-th smallest value is
+definitely contained in `array(left:size(array))`. If `left` is not present,
+the default is 1. This is typically useful if multiple calls to `select` are
+made, because the partial sorting of `array` implies constraints on where we
+need to search.
+
+`right` (optional): shall be a scalar with the same type as `k`. It is an
+`intent(in)` argument. If specified then we assume the k-th smallest value is
+definitely contained in `array(1:right)`. If `right` is not present, the
+default is `size(array)`. This is typically useful if multiple calls to
+`select` are made, because the partial sorting of `array` implies constraints
+on where we need to search.
+
+### Notes
+
+Selection of a single value should have runtime of O(`size(array)`), so it is
+asymptotically faster than sorting `array` entirely. The test program at the
+end of this document shows that is the case.
+
+### Example
+
+```fortran
+    program demo_select
```

Comment:
Here, and in other examples: Why the 4-space indent for the whole example? It seems unnecessary to me, and it will wrap the text on small screens (smartphones).

Reply:
Good point -- there is no reason -- so I removed the outer 4-space-indent in the commit just now.